pkg/controller/sparkapplication/controller.go

```diff
@@ -501,15 +508,25 @@ func (c *Controller) processSingleExecutorStateUpdate(update *executorStateUpdat
 func (c *Controller) updateSparkApplicationStatusWithRetries(
 	original *v1alpha1.SparkApplication,
 	updateFunc func(*v1alpha1.SparkApplicationStatus)) *v1alpha1.SparkApplication {
-	toUpdate := original.DeepCopy()
-
 	var lastUpdateErr error
+	toUpdate := original.DeepCopy()
 	for i := 0; i < maximumUpdateRetries; i++ {
-		updated, err := c.tryUpdateStatus(original, toUpdate, updateFunc)
+
+		glog.V(2).Infof("Trying to update SparkApplication %s", original.Name)
+		// Apply update
+		updateFunc(&toUpdate.Status)
+
+		// Let's keep the old App status if this is not a valid transition.
+		if !isValidDriverStatusTransition(original.Status.AppState.State, toUpdate.Status.AppState.State) {
+			toUpdate.Status.AppState = original.Status.AppState
+		}
+		updated, err := c.tryUpdateStatus(original, toUpdate)
 		if err == nil {
+			c.postMetrics(original, updated)
```

Comment:
We can add a condition around `postMetrics`, i.e., if `updated != nil` we will do `postMetrics`.

Reply:
done